sql: backfill partial indexes

mgartner opened the pull request:
#### sql: backfill partial indexes

This commit adds support for backfilling partial indexes. Adding partial
indexes to tables that already have rows will now correctly backfill the
partial index such that it only indexes rows that satisfy the partial
index.

In order to backfill partial indexes, the backfiller first generates a
typed expression from the partial index predicate string. Then, for each
row in the table, the typed predicate expression is evaluated. If the
result is true, index entries are generated for the partial index.

Fixes #51852 

Release note: None

#### sql: add tests for computed columns that reference other new columns

This commit adds tests for adding computed columns that reference other
new columns added in the same `ALTER TABLE` statement. These cases were
previously untested.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52213)
<!-- Reviewable:end -->

rohany commented on pkg/sql/schemaexpr/expr.go:
```diff
@@ -131,3 +135,84 @@ func ExtractColumnIDs(
+type NameResolver struct {
+	evalCtx    *tree.EvalContext
+	tableID    sqlbase.ID
+	cols       []sqlbase.ColumnDescriptor
```
We want to generally carry slices of `*ColumnDescriptor`, since these are getting expensive to copy around

mgartner replied:
Done.

rohany commented on pkg/sql/logictest/testdata/logic_test/partial_index:
```diff
@@ -455,3 +455,41 @@ SELECT * FROM d@f_b_s_foo_idx WHERE b AND s = 'foo'
+#### Backfill a partial index.
+
+statement ok
+CREATE TABLE e (a INT, b INT)
```
unfortunately, we have two codepaths of how this all runs. One like this, and then one that kicks in when doing schema changes on a new table created in the same transaction. You should pretty much duplicate your tests with those cases.

mgartner replied:
Ahh good to know. Done.

rohany commented on pkg/sql/backfill/backfill.go:
```diff
@@ -468,6 +528,36 @@ func (ib *IndexBackfiller) BuildIndexEntriesChunk(
+		// for.
+		indexesToEncode := ib.added
+		if len(ib.predicates) > 0 {
+			indexesToEncode = make([]sqlbase.IndexDescriptor, 0, len(ib.added))
```
This should be allocated and stored on the index builder so that we don't allocate it per row

rohany commented on pkg/sql/backfill/backfill.go:
```diff
@@ -366,19 +379,65 @@ func (ib *IndexBackfiller) Init(
+
+			// If the index is a partial index, convert and store the predicate
+			// as a TypedExpr.
+			if idx.IsPartial() {
```
Can you take a look at that changes made to the ColumnBackfiller in https://github.com/cockroachdb/cockroach/pull/51865 -- that is the end state for what i imagine these backfillers will look like when there are user defined types in these expressions. Right now, this wouldn't work with user defined types. I'm also OK with it not handling it right now and then opening an issue for it to handle later

mgartner replied:
Nice, looks like that should be pretty straightforward to add. I've added a note in [the issue I'm using to track partial indexes + UDTs](https://github.com/cockroachdb/cockroach/issues/51804). I'll probably go through in one swoop and add a bunch of UDT+partial index tests and fix whatever is broken, if that's ok with you. There's a number of existing bugs at the moment.

rohany replied:
Yup, seems reasonable to me.

rohany commented on pkg/sql/backfill/backfill.go:
```diff
@@ -360,6 +359,15 @@ type IndexBackfiller struct {
+	// indexesToEncode is a list of indexes to encode entries for a given row.
+	// It is a field of IndexBackfiller to avoid allocating a slice for each row
+	// backfilled.
+	indexesToEncode []sqlbase.IndexDescriptor
```
this maybe should be a slice of pointers -- copying descriptors here seems somewhat expensive (since we're going to do it per row)

mgartner replied:
It has to be a `[]sqlbase.IndexDescriptor` unless we change the type signature of `sqlbase.EncodeSecondaryIndexes`. I can go ahead and do that, but it'll be a pretty big change.

rohany replied:
Let's measure it and see (see my comment below). If it does show up as expensive, then I think adding an another function that takes all of the indexes, and then a slice of index ordinals to write would be clean and not affect the API too much.

mgartner replied:
I tested **(A)** regular index backfill on 100k rows on master vs. **(B)** a partial index backfill on 100k rows where every row is in the partial index.

#### A

<img width="986" alt="master" src="https://user-images.githubusercontent.com/1128750/89452352-5d204300-d712-11ea-8e51-17b45b374a92.png">

#### B

<img width="984" alt="partial_index_backfill" src="https://user-images.githubusercontent.com/1128750/89452366-61e4f700-d712-11ea-9584-c3459b1a4d61.png">

mgartner replied:
I've created an issue to track reducing these allocations in the future: https://github.com/cockroachdb/cockroach/issues/52425

rohany commented on pkg/sql/backfill/backfill.go:
```diff
@@ -372,36 +380,135 @@ func (ib *IndexBackfiller) ContainsInvertedIndex() bool {
+	valNeededForCol := ib.initIndexes(desc)
+
+	// Convert any partial index predicate strings into expressions.
+	predicates, predicateRefColIDs, err := schemaexpr.MakePartialIndexExprs(ctx, ib.added, ib.cols, desc, evalCtx, semaCtx)
```
nit: long line

mgartner replied:
Done.

rohany commented on pkg/sql/backfill/backfill.go:
```diff
@@ -482,6 +613,36 @@ func (ib *IndexBackfiller) BuildIndexEntriesChunk(
+				if !idx.IsPartial() {
+					// If the index is not a partial index, all rows should have
+					// an entry.
+					ib.indexesToEncode = append(ib.indexesToEncode, *idx)
```
yeah we shouldn't copy this

rohany commented on pkg/sql/backfill/backfill.go:
```diff
@@ -491,7 +652,7 @@ func (ib *IndexBackfiller) BuildIndexEntriesChunk(
 			ib.evalCtx.Codec,
 			tableDesc.TableDesc(),
-			ib.added,
+			ib.indexesToEncode,
```
oh my goodness its because this takes a slice of nonpointers. That's unfortunate. I'm not sure whats the best thing to do here. I feel like these copies are going to show up in a profile for sure.

mgartner replied:
I'm going to try to change this to take a slice of pointers. Maybe it won't be so bad...

rohany replied:
I think it might be bad :D We use this mainly when we have a table descriptor, so we take that slice of indexes right off of there.

rohany commented on pkg/sql/logictest/testdata/logic_test/computed:
```diff
@@ -700,6 +700,17 @@ SELECT * FROM x ORDER BY a
 3 6 7
 6 12 10
 
+# Multiple ADD COLUMNs can only reference columns that are defined previously.
```
Maybe pull these tests into a separate commit

mgartner replied:
Done.

rohany commented on pkg/sql/logictest/testdata/logic_test/partial_index:
```diff
@@ -4,6 +4,10 @@
 statement ok
 SET experimental_partial_indexes=on
 
+# TODO(mgartner): remove this once enums are fully supported.
```
This TODO is unnecessary

mgartner replied:
Why's that?

rohany commented on pkg/sql/logictest/testdata/logic_test/partial_index:
```diff
@@ -455,3 +459,136 @@ SELECT * FROM d@f_b_s_foo_idx WHERE b AND s = 'foo'
+CREATE TABLE e (a INT, b INT)
+
+statement ok
+INSERT INTO e VALUES (1, 10);
```
just do 1 insert statement with all of the values?

mgartner replied:
Done.

rohany commented on pkg/sql/logictest/testdata/logic_test/partial_index:
```diff
@@ -455,3 +459,136 @@ SELECT * FROM d@f_b_s_foo_idx WHERE b AND s = 'foo'
+5  50
+6  60
+
+#### Backfill a partial index when a new table is created in the same
```
nit: not sure we have a precedent for using comments like this

mgartner replied:
I don't know where I got this from... Fixed.

rohany commented on pkg/sql/logictest/testdata/logic_test/partial_index:
```diff
@@ -455,3 +459,136 @@ SELECT * FROM d@f_b_s_foo_idx WHERE b AND s = 'foo'
+CREATE INDEX a_b_gt_30_idx ON e (a) WHERE b > 30
+
+query II
+SELECT * FROM e@a_b_gt_30_idx WHERE b > 30
```
Is there some way that we can scan the index manually to ensure that there aren't extra KV's? Maybe something like the tests in `schema_changer_test.go`

mgartner replied:
Forcing the index should be the same thing, AFAIK. This test will fail if a row is added to the partial index when it shouldn't be.

mgartner replied:
I've added a comment to make it more clear that this will be a full partial index scan.

rohany reviewed (changes requested):
looking good! mostly small requests here. @lucy-zhang might also want to take a peek at this.

mgartner committed:
sql: backfill partial indexes

This commit adds support for backfilling partial indexes. Adding partial
indexes to tables that already have rows will now correctly backfill the
partial index such that it only indexes rows that satisfy the partial
index.

In order to backfill partial indexes, the backfiller first generates a
typed expression from the partial index predicate string. Then, for each
row in the table, the typed predicate expression is evaluated. If the
result is true, index entries are generated for the partial index.

Release note: None

mgartner committed:
sql: add tests for computed columns that reference other new columns

This commit adds tests for adding computed columns that reference other
new columns added in the same `ALTER TABLE` statement. These cases were
previously untested.

Release note: None

rohany commented on pkg/sql/backfill/backfill.go:
```diff
@@ -372,36 +380,142 @@ func (ib *IndexBackfiller) ContainsInvertedIndex() bool {
+	var predicateRefColIDs sqlbase.TableColSet
+
+	// Install type metadata in the target descriptors, as well as resolve any
+	// user defined types in the column expressions.
```
[nit] any user defined types in partial index expressions

mgartner replied:
Good catch. Done.

rohany commented on pkg/sql/backfill/backfill.go:
```diff
@@ -410,20 +524,38 @@ func (ib *IndexBackfiller) Init(
-	for i := range cols {
-		ib.colIdxMap[cols[i].ID] = i
+
+	ib.types = make([]*types.T, len(ib.cols))
```
Is this used anywhere?

mgartner replied:
Yes here: https://github.com/cockroachdb/cockroach/blob/3b193feb06feac8d4e908a409b1abfad11d57699/pkg/sql/backfill/backfill.go#L481

It's already there on master.

thoszhang reviewed (commented):
The backfiller changes LGTM.

> Reviewed 2 of 16 files at r1, 2 of 4 files at r3, 18 of 18 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52213#-:-ME1YSfQEyM5QnVCMUAL:btldzc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang, @mgartner, and @rohany)



<!-- Sent from Reviewable.io -->

mgartner commented:
Welp just before I thought I passed this level, I think I encountered the final boss, a flaky test failure of the test where an index with a UDT is backfilled within a transaction.

It looks like its a problem with computed columns with UDTs as well on master, so I submitted a ticket here: https://github.com/cockroachdb/cockroach/issues/52539

I want to get this PR in, but I don't want to merge a flaky test.

@rohany any ideas what could be going on? Is `flowCtx.TypeResolverFactory.Descriptors.ReleaseAll(ctx)` not releasing the leases correctly?

rohany commented:
I don't want your PR here to get hung up. Can we just comment out the test here that would flake with a TODO pointing to the new issue?

mgartner commented:
Yes, done.

mgartner commented:
bors r+
